### ISSUE TITLE
view builder hung on shutdown

### ISSUE THREAD

**Issue by bhalevy** [scylladb/scylla#5524]
Seen in https://jenkins.scylladb.com/view/master/job/scylla-master/job/debug-jobs/job/dtest-releng/300/testReport/bootstrap_test/TestBootstrap/start_stop_test_node/
Scylla version 1116700bc976a0c2ae3cd03974d8bf3605ae23f9

Might be related/caused by #5404 

```
Traceback (most recent call last):
  File "/usr/lib64/python2.7/unittest/case.py", line 396, in run
    self.tearDown()
  File "/data/jenkins/workspace/scylla-master/debug-jobs/dtest-releng/scylla-dtest/dtest.py", line 914, in tearDown
    self.cleanUpCluster()
  File "/data/jenkins/workspace/scylla-master/debug-jobs/dtest-releng/scylla-dtest/dtest.py", line 934, in cleanUpCluster
    raise AssertionError("Core file(s) found.{}".format("" if failed else " Marking test as failed."))
AssertionError: Core file(s) found.
Standard Output
node1 is still running. Trying to generate coredump using kill(228, SIGQUIT)...
Moving core file /data/jenkins/workspace/scylla-master/debug-jobs/dtest-releng/scylla-dtest/scylla.228.1577293208.core to /data/jenkins/workspace/scylla-master/debug-jobs/dtest-releng/scylla-dtest/logs-debug.2/1577293285641_bootstrap_test.TestBootstrap.start_stop_test_node/node1-scylla.228.1577293208.core
```

[node1.log](https://jenkins.scylladb.com/view/master/job/scylla-master/job/debug-jobs/job/dtest-releng/300/artifact/logs-debug.2/1577293285641_bootstrap_test.TestBootstrap.start_stop_test_node/node1.log):
```
INFO  2019-12-25 16:59:53,548 [shard 0] init - Shutting down view builder
INFO  2019-12-25 16:59:53,548 [shard 0] view - Stopping view builder
INFO  2019-12-25 16:59:53,548 [shard 1] view - Stopping view builder
```

**Comment by bhalevy** [comment]
@xemul can you please investigate core dump and see if this is explained by #5404?

**Comment by bhalevy** [comment]
Cc @hagitsegev

**Comment by xemul** [comment]
No, that's different. In the #5404 the problem was after "nodetool drain". It cased the migration manager to stop in storage_service::drain() funciton and only after this on regular stop the view builder tried to unregister itself from migration manager.

This bug is different -- I see no signs of regular draining in logs, so the migration_manager instances are alive and view_builder::stop hangs in some other place.

**Comment by xemul** [comment]
In Jenkinks logs:
19:56:27  /build/SCYLLA-RELEASE-FILE:
19:56:27  0.20191224.1116700bc97
In the downloaded scylla-package.tar.gz
$ cat SCYLLA-RELEASE-FILE 
0.20191223.1116700bc97

**Comment by xemul** [comment]
The libs (and maybe the binary itself) seem to be from wrong core, gdb doesn't load them together with eachother and refuses to resolve any symbols in it. Or the core itslef is truncated. Neither this hang is reproduced locally. @hagitsegev can I get remote access on the failing node?

**Comment by xemul** [comment]
The .so files from the package are stripped.

**Comment by bhalevy** [comment]
> The .so files from the package are stripped.

@xemul that's true but it's not the problem...

I reproduced with a more recent package in https://jenkins.scylladb.com/view/master/job/scylla-master/job/debug-jobs/job/dtest-releng/304/artifact/logs-debug.2/
```
Scylla version 666.development-0.20191227.60a851d3a55 with build-id eae225454dfba7ae0c5b0dabd6e1932e025b38b6 starting ...
```

```
$ gdb -c node1-scylla.228.1577611096.core /tmp/scylla-package/libexec/scylla
GNU gdb (GDB) Fedora 8.3.50.20190824-25.fc31
Copyright (C) 2019 Free Software Foundation, Inc.
License GPLv3+: GNU GPL version 3 or later <http://gnu.org/licenses/gpl.html>
This is free software: you are free to change and redistribute it.
There is NO WARRANTY, to the extent permitted by law.
Type "show copying" and "show warranty" for details.
This GDB was configured as "x86_64-redhat-linux-gnu".
Type "show configuration" for configuration details.
For bug reporting instructions, please see:
<http://www.gnu.org/software/gdb/bugs/>.
Find the GDB manual and other documentation resources online at:
    <http://www.gnu.org/software/gdb/documentation/>.

For help, type "help".
Type "apropos word" to search for commands related to "word"...
Reading symbols from /tmp/scylla-package/libexec/scylla...
[New LWP 228]
[New LWP 230]
[New LWP 234]
[New LWP 233]
[New LWP 231]
[New LWP 232]

warning: Could not load shared library symbols for /jenkins/workspace/scylla-master/debug-jobs/dtest-releng/scylla/.ccm/scylla-repository/60a851d3a55f973e65ee71af61516f2750fe4aeb/scylla//opt/scylladb/libreloc/ld.so.
Do you need "set solib-search-path" or "set sysroot"?
[Thread debugging using libthread_db enabled]
Using host libthread_db library "/lib64/libthread_db.so.1".
Core was generated by `/jenkins/workspace/scylla-master/debug-jobs/dtest-releng/scylla/.dtest/dtest-Pg'.
Program terminated with signal SIGQUIT, Quit.
#0  0x00007f6057a857f6 in epoll_pwait () from /usr/lib64/libc-2.30.so
[Current thread is 1 (Thread 0x7f60557b9200 (LWP 228))]
Missing separate debuginfos, use: dnf debuginfo-install boost-date-time-1.69.0-9.fc31.x86_64 boost-filesystem-1.69.0-9.fc31.x86_64 boost-program-options-1.69.0-9.fc31.x86_64 boost-regex-1.69.0-9.fc31.x86_64 boost-system-1.69.0-9.fc31.x86_64 boost-thread-1.69.0-9.fc31.x86_64 c-ares-1.15.0-4.fc31.x86_64 cryptopp-8.2.0-2.fc31.x86_64 fmt-5.3.0-2.fc31.x86_64 glibc-2.30-8.fc31.x86_64 gmp-6.1.2-10.fc31.x86_64 gnutls-3.6.11-1.fc31.x86_64 hwloc-libs-2.0.4-1.fc31.x86_64 jsoncpp-1.9.1-1.fc31.x86_64 libasan-9.2.1-1.fc31.x86_64 libatomic-9.2.1-1.fc31.x86_64 libgcc-9.2.1-1.fc31.x86_64 libgcrypt-1.8.5-1.fc31.x86_64 libgpg-error-1.36-2.fc31.x86_64 libicu-63.2-3.fc31.x86_64 libstdc++-9.2.1-1.fc31.x86_64 libtasn1-4.14-2.fc31.x86_64 libtool-ltdl-2.4.6-31.fc31.x86_64 libubsan-9.2.1-1.fc31.x86_64 libunistring-0.9.10-6.fc31.x86_64 libxcrypt-4.4.10-2.fc31.x86_64 lksctp-tools-1.0.16-12.fc31.x86_64 lua-libs-5.3.5-6.fc31.x86_64 lz4-libs-1.9.1-1.fc31.x86_64 nettle-3.5.1-3.fc31.x86_64 numactl-libs-2.0.12-3.fc31.x86_64 openssl-libs-1.1.1d-2.fc31.x86_64 p11-kit-0.23.16.1-2.fc31.x86_64 protobuf-3.6.1-5.fc31.x86_64 snappy-1.1.7-9.fc31.x86_64 systemd-libs-243.5-1.fc31.x86_64 thrift-0.10.0-19.fc31.x86_64 trousers-lib-0.3.13-13.fc31.x86_64 xz-libs-5.2.4-6.fc31.x86_64 yaml-cpp-0.6.3-1.fc31.x86_64 zlib-1.2.11-20.fc31.x86_64
(gdb) bt
#0  0x00007f6057a857f6 in epoll_pwait () from /usr/lib64/libc-2.30.so
#1  0x000000000acd9f15 in seastar::reactor_backend_epoll::wait_and_process (this=0x604000047450, timeout=-1, active_sigmask=0x633000000820) at /jenkins/workspace/scylla-master/build/scylla/seastar/src/core/reactor_backend.cc:366
#2  0x0000000009fa9a20 in seastar::reactor::wait_and_process (this=0x633000000800, timeout=-1, active_sigmask=0x633000000820) at /jenkins/workspace/scylla-master/build/scylla/seastar/src/core/reactor.cc:930
#3  0x0000000009fe4e5a in seastar::reactor::sleep (this=0x633000000800) at /jenkins/workspace/scylla-master/build/scylla/seastar/src/core/reactor.cc:2761
#4  0x0000000009fe39fd in seastar::reactor::run (this=0x633000000800) at /jenkins/workspace/scylla-master/build/scylla/seastar/src/core/reactor.cc:2728
#5  0x0000000009e2ebd2 in seastar::app_template::run_deprecated(int, char**, std::function<void ()>&&) (this=0x7ffd932e0fe0, ac=28, av=0x7ffd932e1398, func=...) at /jenkins/workspace/scylla-master/build/scylla/seastar/src/core/app-template.cc:199
#6  0x0000000009e2cbf8 in seastar::app_template::run(int, char**, std::function<seastar::future<int> ()>&&) (this=0x7ffd932e0fe0, ac=28, av=0x7ffd932e1398, func=...) at /jenkins/workspace/scylla-master/build/scylla/seastar/src/core/app-template.cc:115
#7  0x000000000050241a in main (ac=28, av=0x7ffd932e1398) at main.cc:463
(gdb) thread 2
[Switching to thread 2 (Thread 0x7f6049f35700 (LWP 230))]
#0  0x00007f6057a857f6 in epoll_pwait () from /usr/lib64/libc-2.30.so
(gdb) bt
#0  0x00007f6057a857f6 in epoll_pwait () from /usr/lib64/libc-2.30.so
#1  0x000000000acd9f15 in seastar::reactor_backend_epoll::wait_and_process (this=0x604000048010, timeout=-1, active_sigmask=0x63300001c820) at /jenkins/workspace/scylla-master/build/scylla/seastar/src/core/reactor_backend.cc:366
#2  0x0000000009fa9a20 in seastar::reactor::wait_and_process (this=0x63300001c800, timeout=-1, active_sigmask=0x63300001c820) at /jenkins/workspace/scylla-master/build/scylla/seastar/src/core/reactor.cc:930
#3  0x0000000009fe4e5a in seastar::reactor::sleep (this=0x63300001c800) at /jenkins/workspace/scylla-master/build/scylla/seastar/src/core/reactor.cc:2761
#4  0x0000000009fe39fd in seastar::reactor::run (this=0x63300001c800) at /jenkins/workspace/scylla-master/build/scylla/seastar/src/core/reactor.cc:2728
#5  0x0000000009ff6566 in seastar::smp::<lambda()>::operator()(void) const (__closure=0x613000002f80) at /jenkins/workspace/scylla-master/build/scylla/seastar/src/core/reactor.cc:3814
#6  0x000000000a04c556 in std::_Function_handler<void(), seastar::smp::configure(boost::program_options::variables_map, seastar::reactor_config)::<lambda()> >::_M_invoke(const std::_Any_data &) (__functor=...) at /usr/include/c++/9/bits/std_function.h:300
#7  0x0000000002a95707 in std::function<void ()>::operator()() const (this=0x6030001026d0) at /usr/include/c++/9/bits/std_function.h:690
#8  0x000000000ad0037f in seastar::posix_thread::start_routine (arg=0x6030001026d0) at /jenkins/workspace/scylla-master/build/scylla/seastar/src/core/posix.cc:52
#9  0x00007f6058d534e2 in start_thread () from /usr/lib64/libpthread-2.30.so
#10 0x00007f6057a85693 in clone () from /usr/lib64/libc-2.30.so
```

**Comment by bhalevy** [comment]
Plenty more examples in https://jenkins.scylladb.com/view/master/job/scylla-master/job/dtest-debug/353/testReport/

**Comment by xemul** [comment]
In the core there's a thread_context with view_builder::start callable on boards. Apparently the start hasn't finished yet and view_builder::stop waits for it.

**Comment by xemul** [comment]
For the record, thread_contexts out threre:
shard1:
view12view_builder5start
view21view_update_generator5start
hints14space_watchdog5start
shard0:
view12view_builder5start
view21view_update_generator5start
hints14space_watchdog5start
service26view_update_backlog_broker5start
main

**Comment by xemul** [comment]
There's also a bunch of timers with expiration time-stamp between 3622513s and 3626115s which is in 1970 %) Likely I'm missing some offset.

**Comment by xemul** [comment]
There are two reactors with _lowres_next_timeout set to 3622632s and 3622632s.
The timers in question are having expiration values between 3622512s and 3622632744s but they are not run by the reactor. One of the times is sleeper in view_builder::start.

**Comment by xemul** [comment]
Both reactors sit in the epoll_pwait()

**Comment by xemul** [comment]
BTW, there are 726 "stuck" timers out there.

**Comment by xemul** [comment]
It would be nice to have the stuck node at hands to explore not only the core, but also the process attributes, such as pending/blocked sigmasks, states of epoll/event/timerfd-s etc.

**Comment by xemul** [comment]
Reactors' pending task queues are all empty.

**Comment by xemul** [comment]
Threads:
1, 2. in reactor loop's epoll_pwait()
3, 4. task_quota_timer_thread_fn -> read from fd
5, 6. thread_pool::work -> read from eventfd

**Comment by xemul** [comment]
@hagitsegev , what Linux distro (in particular -- which kernel) is used in Jenkins slaves during this test?

**Comment by xemul** [comment]
The reactor._signal._pending_signals is 0 for both shards.

**Comment by xemul** [comment]
shard1 thread sitting in epoll_pwait has signal 34 pending, but blocked. Signal 34 is SIGRTMIN, the one that's used by seastar to deliver timer events :\

**Comment by xemul** [comment]
@hagitsegev , what Linux distro (in particular -- which kernel) is used in Jenkins slaves during this test?

**Comment by xemul** [comment]
Both shards have _timer_enabled = true

**Comment by xemul** [comment]
Shard with blocked timer signal has _timers._last = 0 and _timers._next = 3622511s.

**Comment by xemul** [comment]
There's one suspicious place in reactor::signal_pollfn::try_enter_interrupt_mode(). If it finds a pending signal and handles it, in case the handler adds a timer, then the timer will be set-up and the signal unblocked, but the exit_interrupt_mode() will re-block the signal back again. Should it race once, the timer signal will be blocked forever, as enabling the alarm_signal() happens only once in arm_highres_timer() due to _timer_enabled check.

**Comment by xemul** [comment]
This is wierd. The vector of pollers looks like this:

$59 = {_vptr.pollfn = 0x15377a58 <vtable for seastar::reactor::execution_stage_pollfn+16>}
$60 = {_vptr.pollfn = 0x15377b58 <vtable for seastar::reactor::aio_batch_submit_pollfn+16>}
$61 = {_vptr.pollfn = 0x15377b18 <vtable for seastar::reactor::drain_cross_cpu_freelist_pollfn+16>}
$62 = {_vptr.pollfn = 0x15377a98 <vtable for seastar::reactor::smp_pollfn+16>}
$63 = {_vptr.pollfn = 0x15377a18 <vtable for seastar::reactor::syscall_pollfn+16>}
$64 = {_vptr.pollfn = 0x15377c18 <vtable for seastar::reactor::io_pollfn+16>}
$65 = {_vptr.pollfn = 0x15377b98 <vtable for seastar::reactor::batch_flush_pollfn+16>}
$66 = {_vptr.pollfn = 0x15377bd8 <vtable for seastar::reactor::signal_pollfn+16>}
$67 = {_vptr.pollfn = 0x15377ad8 <vtable for seastar::reactor::lowres_timer_pollfn+16>}
$68 = {_vptr.pollfn = 0x153779d8 <vtable for seastar::reactor::epoll_pollfn+16>}

note the vars $66 and $67. The comment in reactor::run explicitly warns about this bad order:

    // expire_lowres_timers must be before sig_poller, because lowres_timer_pollfn
    // may arm the first highres timer, which can add a new signal to be registerd. If the order
    // is reversed, then signal_pollfn::exit_interrupt_mode() can re-block the timer signal.

and this is what we've somehow stepped into.

**Comment by avikivity** [comment]
> There's also a bunch of timers with expiration time-stamp between 3622513s and 3626115s which is in 1970 %) Likely I'm missing some offset.

Those timers use std::steady_clock as a base, whose epoch is the boot time. Those numbers are about 100 hours = 4 days, which sounds right.

Compare to lowres_clock_impl::_steady_now to see if they should have expired or not.

**Comment by xemul** [comment]
The _steady_now seems to be 3622632 on shard 0, but the problem with timers must be different -- I've described that the _pollers vector is not ordered the way it's expected to be.

**Comment by avikivity** [comment]
I think I see what happens. Pollers are registered indirectly via a registration_task, but tasks can be reordered in the run queue.

Is this a debug build? the debug build randomizes the task queue, just because.

**Comment by avikivity** [comment]
Yes, the link says "debug-jobs"

**Comment by avikivity** [comment]
See `SEASTAR_SHUFFLE_TASK_QUEUE`

**Comment by avikivity** [comment]
So, the problem only happens on debug builds. That's how such a major bug in the infrastructure escaped detection for so long, it requires many days to reproduce and debug builds rarely run so long.

**Comment by xemul** [comment]
@avikivity , any reason why SIGRTMIN (a.k.a. alarm_signal()) handler is set in arm_highres_timer(), i.e. by the time the first timer is queued, but not (likely much) earlier?

**Comment by avikivity** [comment]
It happened in scylladb/scylla@5aea50a8a4f46316bb5015a79989c3281d4110e7. I'm guessing the reason is that reactor_backend construction is too early, but maybe it's not.

**Comment by avikivity** [comment]
Aged out.
